waspc/data/Generator/templates/react-app/src/webSocket.ts

```diff
@@ -0,0 +1,76 @@
+{{={= =}=}}
+
+import { useState, useEffect } from 'react'
+import { io, Socket } from 'socket.io-client'
+import { DefaultEventsMap, EventsMap } from '@socket.io/component-emitter'
+
+import { getAuthToken, events as apiEvents, EventType } from './api'
+import config from './config'
+
+// TODO: In the future, it would be nice if users could pass more
+// options to `io`, likely via some `configFn`.
+export const socket = io(config.apiUrl, { autoConnect: {= autoConnect =} })
+
+function refreshAuthToken() {  
+  // NOTE: When we figure out how `auth: true` works for Operations, we should
+  // mirror that behavior here for WebSockets. Ref: https://github.com/wasp-lang/wasp/issues/1133
+  socket.auth = {
+    token: getAuthToken()
+  }
+
+  if (socket.connected) {
+    socket.disconnect()
+    socket.connect()
+  }
+}
+
+refreshAuthToken()
+apiEvents.on(EventType.SET_AUTH_TOKEN, refreshAuthToken)
+apiEvents.on(EventType.CLEAR_AUTH_TOKEN, refreshAuthToken)
+
+export function useSocket<
+  ServerToClientEvents extends EventsMap = DefaultEventsMap,
+  ClientToServerEvents extends EventsMap = DefaultEventsMap
+>(): {
+  socket: Socket<ServerToClientEvents, ClientToServerEvents>;
+  isConnected: boolean;
+} {
+  const [isConnected, setIsConnected] = useState(socket.connected)
+
+  useEffect(() => {
```

Comment:
I think there is an opportunity to use context here:
- define a `WebSocketContext`
- wrap our app in the `WebSocketProvider`
    ```jsx
    <WebSocketProvider value={{ scoket, isConnected }}>
      {children}
    <WebSocketProvider>
    ```
- keep the state in the `WebSocketContext`
- `useSocket` gets it's socket from there (kinda like global `useState`)
    ```ts
    const { socket } = useContext(WebSocketContext);
    ```
- then ... `useSocketListener` can use `useSocket` under the hood and the user doesn't even need to think about providing the `socket` to `useSocketListener`

TLDR: we maybe can reduce the API of `useSocketListener` to `event` + `cb`

I can try to rewrite this to that form and commit if it works well.

Reply:
Sounds interesting!

Btw I think this could work even without us using `context` though, right? I mean we can do all this right now, by using `useSocket` in `useSocketListener`? What do we get with `context` that we don't have now? I am not against `context`, just wondering.